File: examples/chapter8.rb
require_relative "../lib/rayz/world"
require_relative "../lib/rayz/camera"
require_relative "../lib/rayz/point_light"
require_relative "../lib/rayz/point"
require_relative "../lib/rayz/color"
require_relative "../lib/rayz/sphere"
require_relative "../lib/rayz/plane"
require_relative "../lib/rayz/transformations"
require_relative "../lib/rayz/material"
require_relative "../lib/rayz/stripe_pattern"
require_relative "../lib/rayz/checkers_pattern"
require_relative "../lib/rayz/gradient_pattern"
require_relative "../lib/rayz/ring_pattern"

module Rayz
  class Chapter8
    def self.run
      puts "Chapter 8: Shadows (Patterns and Planes)"

      # Create a world
      world = World.new

      # Add light source
      world.light = PointLight.new(
        position: Point.new(x: -10, y: 10, z: -10),
        intensity: Color.new(red: 1, green: 1, blue: 1)
      )

      # Create floor plane with checkers pattern
      floor = Plane.new
      floor.material.pattern = CheckersPattern.new(
        a: Color.new(red: 1, green: 1, blue: 1),
        b: Color.new(red: 0.2, green: 0.2, blue: 0.2)
      )
      floor.material.specular = 0
      world.objects << floor

      # Create back wall plane with gradient pattern
      back_wall = Plane.new
      back_wall.transform = Transformations.rotation_x(radians: Math::PI / 2) *
        Transformations.translation(x: 0, y: 0, z: 5)
      back_wall.material.pattern = GradientPattern.new(
        Color.new(red: 0.5, green: 0.7, blue: 1),
        Color.new(red: 0.1, green: 0.1, blue: 0.3)
      )
      back_wall.material.pattern.transform = Transformations.rotation_z(radians: Math::PI / 2) *
        Transformations.scaling(x: 2, y: 2, z: 2)
      back_wall.material.specular = 0
      world.objects << back_wall

      # Create middle sphere with ring pattern
      middle = Sphere.new
      middle.transform = Transformations.translation(x: -0.5, y: 1, z: 0.5)
      middle.material.pattern = RingPattern.new(
        Color.new(red: 0.1, green: 1, blue: 0.5),
        Color.new(red: 0.9, green: 0.1, blue: 0.9)
      )
      middle.material.pattern.transform = Transformations.scaling(x: 0.2, y: 0.2, z: 0.2)
      middle.material.diffuse = 0.7
      middle.material.specular = 0.3
      world.objects << middle

      # Create right sphere with stripe pattern
      right = Sphere.new
      right.transform = Transformations.translation(x: 1.5, y: 0.5, z: -0.5) *
        Transformations.scaling(x: 0.5, y: 0.5, z: 0.5)
      right.material.pattern = StripePattern.new(
        Color.new(red: 1, green: 0.2, blue: 0.2),
        Color.new(red: 1, green: 1, blue: 0.2)
      )
      right.material.pattern.transform = Transformations.scaling(x: 0.2, y: 0.2, z: 0.2) *
        Transformations.rotation_z(radians: Math::PI / 4)
      right.material.diffuse = 0.7
      right.material.specular = 0.3
      world.objects << right

      # Create left sphere with gradient pattern
      left = Sphere.new
      left.transform = Transformations.translation(x: -1.5, y: 0.33, z: -0.75) *
        Transformations.scaling(x: 0.33, y: 0.33, z: 0.33)
      left.material.pattern = GradientPattern.new(
        Color.new(red: 1, green: 0.8, blue: 0.1),
        Color.new(red: 0.1, green: 0.2, blue: 1)
      )
      left.material.pattern.transform = Transformations.translation(x: -1, y: 0, z: 0) *
        Transformations.scaling(x: 2, y: 2, z: 2)
      left.material.diffuse = 0.7
      left.material.specular = 0.3
      world.objects << left

      # Set up camera
      camera = Camera.new(hsize: 400, vsize: 200, field_of_view: Math::PI / 3)
      camera.transform = Transformations.view_transform(
        from: Point.new(x: 0, y: 1.5, z: -5),
        to: Point.new(x: 0, y: 1, z: 0),
        up: Vector.new(x: 0, y: 1, z: 0)
      )

      puts "Rendering scene with patterns and planes (400x200 pixels)..."
      puts "Features:"
      puts "  - Floor: Checkers pattern on an infinite plane"
      puts "  - Back wall: Gradient pattern"
      puts "  - Middle sphere: Ring pattern"
      puts "  - Right sphere: Stripe pattern (rotated)"
      puts "  - Left sphere: Gradient pattern"
      puts "This may take a minute..."

      # Render the scene
      canvas = camera.render(world)

      # Write to file
      File.write("examples/chapter8.ppm", canvas.to_ppm)
      puts "Scene rendered to chapter8.ppm"
      puts "\n" + ("=" * 60) + "\n"
    end
  end
end

File: python/examples/chapter8.py
"""Chapter 8: Patterns and Planes — surfaces with patterns on infinite planes."""

from __future__ import annotations

import math
import os

from rayz.camera import Camera
from rayz.color import Color
from rayz.pattern import checkers_pattern, gradient_pattern, ring_pattern, stripe_pattern
from rayz.plane import Plane
from rayz.point_light import PointLight
from rayz.sphere import Sphere
from rayz.transformations import rotation_x, rotation_z, scaling, translation, view_transform
from rayz.tuple import Point, Vector
from rayz.world import World


def run() -> None:
    print("\n=== Chapter 8: Patterns and Planes ===\n")

    world = World()
    world.light = PointLight(Point(-10, 10, -10), Color(1, 1, 1))

    floor = Plane()
    floor.material.pattern = checkers_pattern(Color(1, 1, 1), Color(0.2, 0.2, 0.2))
    floor.material.specular = 0
    world.objects.append(floor)

    back_wall = Plane()
    back_wall.set_transform(rotation_x(math.pi / 2) * translation(0, 0, 5))
    back_wall.material.pattern = gradient_pattern(Color(0.5, 0.7, 1), Color(0.1, 0.1, 0.3))
    back_wall.material.pattern.set_transform(rotation_z(math.pi / 2) * scaling(2, 2, 2))
    back_wall.material.specular = 0
    world.objects.append(back_wall)

    middle = Sphere()
    middle.set_transform(translation(-0.5, 1, 0.5))
    middle.material.pattern = ring_pattern(Color(0.1, 1, 0.5), Color(0.9, 0.1, 0.9))
    middle.material.pattern.set_transform(scaling(0.2, 0.2, 0.2))
    middle.material.diffuse = 0.7
    middle.material.specular = 0.3
    world.objects.append(middle)

    right = Sphere()
    right.set_transform(translation(1.5, 0.5, -0.5) * scaling(0.5, 0.5, 0.5))
    right.material.pattern = stripe_pattern(Color(1, 0.2, 0.2), Color(1, 1, 0.2))
    right.material.pattern.set_transform(scaling(0.2, 0.2, 0.2) * rotation_z(math.pi / 4))
    right.material.diffuse = 0.7
    right.material.specular = 0.3
    world.objects.append(right)

    left = Sphere()
    left.set_transform(translation(-1.5, 0.33, -0.75) * scaling(0.33, 0.33, 0.33))
    left.material.pattern = gradient_pattern(Color(1, 0.8, 0.1), Color(0.1, 0.2, 1))
    left.material.pattern.set_transform(translation(-1, 0, 0) * scaling(2, 2, 2))
    left.material.diffuse = 0.7
    left.material.specular = 0.3
    world.objects.append(left)

    camera = Camera(200, 100, math.pi / 3)
    camera.transform = view_transform(Point(0, 1.5, -5), Point(0, 1, 0), Vector(0, 1, 0))

    print("Rendering 200x100 scene with patterns on planes...")
    canvas = camera.render(world)

    out_path = os.path.join(os.path.dirname(__file__), "chapter8.ppm")
    print(f"Writing {out_path}...", end="", flush=True)
    with open(out_path, "w") as f:
        f.write(canvas.to_ppm())
    print(" done.")
    print("Patterns on planes rendered.")
    print("\n" + "=" * 60 + "\n")


if __name__ == "__main__":
    run()
